### ISSUE TITLE
key as Buffer and zero-copy?

### ISSUE THREAD

**Issue by Llorx** [DoctorEvidence/lmdb-js#110]
I have my own serialization/deserialization method to a Buffer, for both keys and values.

I can use it with levelDB and now trying to use it with lmdb without success:
```
let myDB = open({
    path: dbPath,
    encoding: "binary",
    keyEncoding: "binary"
});

let k = Buffer.from("key");
let v = Buffer.from("value");

await myDb.put(k, v);
for (let { key, value } of myDB.getRange({})) {
    // Error. key is not a Buffer, but an array of 3 elements
}
```

Also, I've seen the `writeKey` and `readKey` methods, but I'm not sure as they involve buffer copying.

Is there a way to use Buffers as key/values? Also there's a way to make them zero-copy? Like node-lmbd (but I like this one more, that's why I'm asking).

Thank you.

**Comment by kriszyp** [comment]
> Error. key is not a Buffer, but an array of 3 elements
> Is there a way to use Buffers as key/values?

Yes, you should be able to use buffers as keys and values (and they should roundtrip). I am not able to reproduce this issue with keys coming back as non-buffers. When I run your code, I get a Buffer as the key and value (the Buffer _is_ an array-like structure with three elements corresponding to the three bytes). Are you running this on 2.0.x? Are the values buffers? In trying to reproduce this, I did notice that using `getRange({})` with binary keys without `start` and `end` causes an error, since there isn't a default conversion of undefined to a buffer/binary. I will fix that (let me know if you want it published on a certain version line).

> Also there's a way to make them zero-copy?

I assume you are asking about this for performance. One of the hard-won lessons of endlessly benchmarking and profiling lmdb-js and node-lmdb is the surprising and unintuitive fact that _creating_ buffers (creating new TypedArray+ArrayBuffer instances) is _vastly_ more expensive than _copying_ buffers (except perhaps for the case of very very large buffers). node-lmdb's zero copy get functions (that create a new buffer instance pointing to existing memory) are actually much slower than alternate methods that utilize buffer reuse (copying data in an existing/reusable buffer). Consequently, most lmdb-js's functions are built around fast and efficient buffer reuse, as this really provides far more performance boosting than any zero-copy techniques that I have seen in all my testing/working.

> Also, I've seen the writeKey and readKey methods

writeKey and readKey functions, as well getBinaryFast, allow for serialization and deserialization of keys and values using lmdb-js's internal buffer reuse techniques, and do typically provide valuable performance benefits. Let me know if you have questions on how to use them.

**Comment by Llorx** [comment]
I'm with `2.1.0-beta1` version.

The error is when using `getRange`, yes. I need to access the entire database and is the only way I can think of. I have not tried to just `put` and `get` a specific key. My question was more like "Can I use `getRange` when using binary Buffers as keys?".

The `write/readKey` question was just in case maybe is the way to go for `getRange` to work, but because I saw buffer copy I was not sure if was a performant way or just a little helper for weird situations, but with less performance.

So yes, the problem is just the `getRange` when using Buffer keys and without a start nor end, nothing more. Is the only way to iterate though the entire database without knowing the start or end keys?

Thank you.

**Comment by Llorx** [comment]
Ooohh I see. I though that start and end should be an existing key (which I don't know haha). Thank you. Will go with that by now.

**Comment by kriszyp** [comment]
Ok, I published v2.1.0-beta2 with the fix to allow getRange({}) with binary keys (to traverse whole db).
